### PR TITLE
[Xamarin.Android.Build.Tasks] reduce task time

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
This commit does 3 changes to ConvertResourceCase, in order to
drastically reduces execution times. On my sample, execution times are
less than half of the original times:

- simplify the linq query used to get the `xmls` files (thx @jonathanpeppers)
- compute `reosurcedictionaries` out of the loop
- do not create a temp file as UpdateXmlResource already creates a temp
file. Change the return value of UpdateXmlResource to indicate if the
file has been changed or not.

### CODE REVIEW

**Comment by jonathanpeppers** [comment]
I think the removal of the temp files are breaking something:
```
  EXEC : 1) error : Xamarin.Android.Build.Tests.AndroidUpdateResourcesTest.Check9PatchFilesAreProcessed(True) [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets]
  MainActivity.cs(22,36,22,40): error CS0117: 'Resource.Layout' does not contain a definition for 'Main' [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/TestDebug/temp/Check9PatchFilesAreProcessed_True/Application1/UnnamedProject.csproj] [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets]
  MainActivity.cs(26,54,26,62): error CS0117: 'Resource.Id' does not contain a definition for 'myButton' [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/TestDebug/temp/Check9PatchFilesAreProcessed_True/Application1/UnnamedProject.csproj] [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets]
  MainActivity.cs(22,36,22,40): error CS0117: 'Resource.Layout' does not contain a definition for 'Main' [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/TestDebug/temp/Check9PatchFilesAreProcessed_True/Application1/UnnamedProject.csproj] [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets]
  MainActivity.cs(26,54,26,62): error CS0117: 'Resource.Id' does not contain a definition for 'myButton' [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/bin/TestDebug/temp/Check9PatchFilesAreProcessed_True/Application1/UnnamedProject.csproj] [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets]
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets(38,5): error MSB3073: The command "mono packages/NUnit.ConsoleRunner.3.7.0/tools/nunit3-console.exe  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/../../bin/TestDebug/Xamarin.Android.Build.Tests.dll  --labels=All --result="TestResult-Xamarin.Android.Build.Tests.xml;format=nunit2" --output="bin/TestDebug/TestOutput-Xamarin.Android.Build.Tests.txt"" exited with code 1.

    128 Warning(s)
    6 Error(s)
```

Will look into it tomorrow.

**Comment by dellis1972** [comment]
@StephaneDelcroix can we revert just the Temp file changes and leave the Linq and directory changes in? We can look at the temp file Changes in a different PR.

**Comment by pierceboggan** [comment]
@StephaneDelcroix Could you share some binlogs with the speedup we see with this change?

**Comment by StephaneDelcroix** [comment]
@dellis1972 sure I could cut that in part, but the biggest gain in speed _is_ from avoiding the double-temp file situation

@pierceboggan sure

**Comment by dellis1972** [comment]
@StephaneDelcroix yes, but it looks like the temp file changes broke a few tests. I would like to get the linq changes in asap, then we can focus a new PR on those temp changes. Otherwise the PR will be held up until it is green. Alternatively, fix the problem that the test is highlighting, but I suspect that will take a bit more time.

**Comment by dellis1972** [comment]
build

**Comment by dellis1972** [comment]
build

File: src/Xamarin.Android.Build.Tasks/Tasks/ConvertResourcesCases.cs [hunk 1]
```diff
@@ -53,18 +53,25 @@ void FixupResources (ITaskItem item, Dictionary<string, string> acwMap)
...
+				xmls.Add (file);
+			}
 
 			var lastUpdate = DateTime.MinValue;
 			if (!string.IsNullOrEmpty (AndroidConversionFlagFile) && File.Exists (AndroidConversionFlagFile)) {
 				lastUpdate = File.GetLastWriteTimeUtc (AndroidConversionFlagFile);
 			}
 			Log.LogDebugMessage ("  AndroidConversionFlagFile modified: {0}", lastUpdate);
+			
+			var resourcedirectories = ResourceDirectories.Where (s => s != item).Select(s => s.ItemSpec).ToArray();
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tasks/ConvertResourcesCases.cs]: @StephaneDelcroix I think too much got taken out.

Something should use this variable?

Comment (StephaneDelcroix) [reply src/Xamarin.Android.Build.Tasks/Tasks/ConvertResourcesCases.cs]: forgot to force push. fixed

**Pull request merged**